src/audio/module_adapter/module_adapter.c

```diff
@@ -1072,13 +1072,27 @@ static int module_adapter_copy_dp_queues(struct comp_dev *dev)
 	list_for_item(blist, &dev->bsink_list) {
 		/* output - we need to copy data from dp_queue (as source)
 		 * to audio_stream (as sink)
+		 *
+		 * a trick is needed there as a workaround
+		 * DP may produce a huge chunk of output data (i.e. 10 LL
+		 * cycles), and the following module should be able to consume it in 1 cycle chunks
+		 *
+		 * unfortunately some modules are not prepared to work when there's more than
+		 * 1 data portion available in the buffer and are draining buffers with data loss
+		 *
+		 * a workaround: copy only the following module's IBS in each LL cycle
+		 *
+		 * required fix: all modules using sink/src interface must be aware to
+		 * process only data they need, not forcefully draining a buffer
 		 */
 		assert(dp_queue);
 		struct comp_buffer *buffer =
 				container_of(blist, struct comp_buffer, source_list);
 		struct sof_sink *data_sink = audio_stream_get_sink(&buffer->stream);
+		struct sof_source *following_mod_data_source =
+				audio_stream_get_source(&buffer->stream);
 		struct sof_source *data_src = dp_queue_get_source(dp_queue);
-		uint32_t to_copy = MIN(sink_get_free_size(data_sink),
+		uint32_t to_copy = MIN(source_get_min_available(following_mod_data_source),
```

Comment:
shouldn't we fix offending LL modules one by one - as soon as you generate respective topologies?

Reply:
absolutely we do
But at the moment we need a working WA for google AEC topology - and we need it ASAP

Reply:
ok, since this is a temp WA, we probably want to create an issue to track and link it to the comments here. We can then check all the other modules as we build topologies and remove the WA.
I think it also make sense to put the WA in an ```ifdef MODULE_WA_FOR_BLAH``` so that its easy for anyone to fix all modules and cleanly remove (and test module updates)..

Reply:
This is at least in dp_copy_queues() function, so specific to DP-LL interface and the comment explains what is done here, so could be worse.

Reply:
@lgirdwood I don't think it is neccessary
dp_copy_queues in fact is a kind of WA itself. It is a double-buffer copying from LL buffers to pipeline 2.0 sink/src interface. It makes a DP module to look from the other components point of view (almost) exactly as a "normal" LL

Once all modules use sink/src and communication between them is in pipeline 2.0 manner (I'm preparng an RFC for this), dp_copy_queues will be removed in whole (well, hopefully together with module adapter)

Reply:
@marcinszkudlinski ok, and we should be able to track this in RFC GH issue. i.e. we can mention all the PRs - sound like we target v2.9 for this too.